### ISSUE TITLE
Compilation issue with msvc 2015 on 32 bits target

### ISSUE THREAD

**Issue by fabienhe** [agenium-scale/nsimd#76]
Hello,

We are facing a compilation issue using nsimd on 32 bit target (VS2015 toolset v140_xp) : 

> 1>D:\Projets\DevSources\platform_sw_nsimd\sw_pico\..\common\misc_lib\nsimd\include\nsimd/nsimd.h(419): error C3861: '__popcnt64' : identificateur introuvable
1>D:\Projets\DevSources\platform_sw_nsimd\sw_pico\..\common\misc_lib\nsimd\include\nsimd/x86/sse2/storeu.h(47): error C2719: 'a1' : le paramètre formel avec l'alignement demandé 16 ne sera pas aligné
1>D:\Projets\DevSources\platform_sw_nsimd\sw_pico\..\common\misc_lib\nsimd\include\nsimd/x86/sse2/storeu.h(75): error C2719: 'a1' : le paramètre formel avec l'alignement demandé 16 ne sera pas aligné
1>D:\Projets\DevSources\platform_sw_nsimd\sw_pico\..\common\misc_lib\nsimd\include\nsimd/x86/sse42/storeu.h(47): error C2719: 'a1' : le paramètre formel avec l'alignement demandé 16 ne sera pas aligné
1>D:\Projets\DevSources\platform_sw_nsimd\sw_pico\..\common\misc_lib\nsimd\include\nsimd/x86/sse42/storeu.h(75): error C2719: 'a1' : le paramètre formel avec l'alignement demandé 16 ne sera pas aligné
1>D:\Projets\DevSources\platform_sw_nsimd\sw_pico\..\common\misc_lib\nsimd\include\nsimd/x86/sse2/eq.h(47): error C2719: 'a0' : le paramètre formel avec l'alignement demandé 16 ne sera pas aligné
1>D:\Projets\DevSources\platform_sw_nsimd\sw_pico\..\common\misc_lib\nsimd\include\nsimd/x86/sse2/eq.h(47): error C2719: 'a1' : le paramètre formel avec l'alignement demandé 16 ne sera pas aligné
1>D:\Projets\DevSources\platform_sw_nsimd\sw_pico\..\common\misc_lib\nsimd\include\nsimd/x86/sse2/eq.h(56): error C2719: 'a0' : le paramètre formel avec l'alignement demandé 16 ne sera pas aligné
1>D:\Projets\DevSources\platform_sw_nsimd\sw_pico\..\common\misc_lib\nsimd\include\nsimd/x86/sse2/eq.h(56): error C2719: 'a1' : le paramètre formel avec l'alignement demandé 16 ne sera pas aligné

the first line (error C3861: '__popcnt64' : identificateur introuvable) make me think that it has never been tested with such target

Thanks in advance

**Comment by gquintin** [comment]
Hi @fabienhe, Indeed we have never tested NSIMD on 32-bits Windows/MSVC. From what I see in [Microsoft's documentation](https://docs.microsoft.com/en-us/cpp/intrinsics/popcnt16-popcnt-popcnt64?view=msvc-160) `__popcnt64` is not available in 32-bits mode. I will fix this and come back to you.

**Comment by gquintin** [comment]
@fabienhe I was looking at the other errors you have (C2719). Have you something special in your code on the SIMD types? I will double check but we have not placed any alignment constraints on SIMD types.

**Comment by gquintin** [comment]
@fabienhe I have fixed the issue in the spmd branch that I will merge in the coming days. I let the issue opened as I have other errors on Intel intrinsics not available in 32-bits mode.

**Comment by fabienhe** [comment]
Thanks,

About issues (C2719) it seems that compiler does not support alignment for parameter value, parameters should be ref or pointer...

**Comment by gquintin** [comment]
@fabienhe I have fixed the issues related to MSVC 32-bits. I have compiled with MSVC 2019 and I do not have the C2719 error. I cannot use references as it is not supported in plain C. Passing pointers seem a bad idea at first glance. I mark this issue as resolved for the next merge. In the meantime I will look at godbolt for the C2719 issue.

**Comment by fabienhe** [comment]
Ok thank you, here a similar issue :

https://bugzilla.mozilla.org/show_bug.cgi?id=1442554

**Comment by gquintin** [comment]
Hi @fabienhe,

I installed MSVC 2015 on one of our Windows servers. The error seems to come from Microsoft header themselves: `C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\[ex]mmintrin.h`:

```c++
typedef union __declspec(intrin_type) __declspec(align(16)) __m128 { ...
typedef union __declspec(intrin_type) __declspec(align(16)) __m128i { ...
typedef union __declspec(intrin_type) __declspec(align(16)) __m128d { ...
```

They require a 16-bytes alignment on SIMD types. I commented out the `__declspec(align(16))` and only functions involving 4 or more SIMD arguments were posing problems, i.e. only the `store4[au]`. Searching a bit more I came accross [this page](https://devblogs.microsoft.com/cppblog/introducing-vector-calling-convention/) explaining that problem and that problems begin with the fourth argument. Using the `/Gv` compiler switch seems to solve the problem. Other problems appear then with MSVC 2015. I am investigating them now.

**Comment by gquintin** [comment]
@fabienhe, compiling with `/Gv` seems to work, I have fixed the other issues. They all had the same origin. The fact that `__cplusplus` is not reported correctly by MSVC and for some versions `_MSVC_LANG` has to be used:

- <https://docs.microsoft.com/en-us/cpp/build/reference/zc-cplusplus?view=msvc-160>
- <https://docs.microsoft.com/en-us/cpp/preprocessor/predefined-macros?view=msvc-160>

Is something like `/Gv` ok for your code?

**Comment by gquintin** [comment]
Ok I have done the "proper" thing which is use the `__vectorcall` modifier on NSIMD functions so that you don't have to use the `/Gv` compiler switch which has an effect on the whole program. The use of `/Gv` is the quick'n'dirty way. This issue should be resolved by commit 018b487d827c119d906ef78dd0d70f3a5997d852 of the spmd branch.

I am keeping this issue opened for now to have your feedback. Hope it works for you. Note that if all goes well I will merge tomorrow and release a v2 which will include all the fixes including this one.

**Comment by fabienhe** [comment]
Hi @gquintin,

Yes we also have tried the `__vectorcall` patch on the simd library and it seems to work (build ended without errors). So I think the issue would be closed for us.

Thank you
